libyang: upgrade to v3 step 1 -- import libyang3 / libyang3-py3, remove stale versions

bradh352 opened the pull request:
#### Why I did it

Upgrade to libyang3

See master tracking ticket #22385 

Fixes https://github.com/sonic-net/sonic-buildimage/issues/22385

##### Work item tracking
N/A

#### How I did it

* Remove unused libyang 1.0.184
* Import libyang 3.12.2
* Import libyang-python 3.1.0
* Migrate FRR to use libyang 3.12.2
* Update yang model test cases in preparation of libyang3 (will make the next PR easier to understand)

**Important**: Please review each commit (including commit message) individually. Looking at the patch-set as a whole may cause confusion.

#### How to verify it

See that the build still succeeds and existing test cases still pass

#### Which release branch to backport (provide reason below if selected)

N/A

#### Tested branch (Please provide the tested image version)

master as of 20250613

#### Description for the changelog
libyang: upgrade to v3 step 1

#### Link to config_db schema for YANG module changes
N/A

#### A picture of a cute animal (not mandatory but encouraged)

Signed-off-by: Brad House (@bradh352)

bradh352 committed:
libyang3-py3: libyang3 uses a python cffi bindings

libyang 1.0.73 provided python bindings via swig.  Those have
been removed and libyang3 now requires the use of the cffi bindings
which are in a separate repo.

The binding names themselves are actually different ('yang' vs 'libyang')
so both should be installable simultaneously.

This is built as a debian package rather than a wheel since it contains
binary elements.

bradh352 committed:
FRR: upgrade to libyang3

We omit the _UNINSTALLS because it messes up dependency tracking with
other packages that depend on libyang3

bradh352 committed:
libyang3-py3: patch to support json int/bool as strings

https://github.com/CESNET/libyang-python/pull/134

needed for backwards compatibility for existing configurations.

bradh352 committed:
sonic-yang-models: add dependency on libyang3-py3 (unused)

We need to force libyang3-py3 to be built and tested, and confirm
it can properly co-exist with libyang1 and its python bindings.

Its currently not used, but will be in a followup.

bradh352 committed:
sonic-yang-models: Update test cases in preparation for libyang3

This mostly simplifies the existing test cases and standardizes
behavior where they may be incompatible with libyang3.  The main goal
is to have this in place before the migration to libyang3 to make it
easier to evaluate the actual code review for porting aspects for libyang3.

Convert some yang model test cases that use `eStr` for failure matching
to use `eStrKey` instead to standardize expected failure matching so
when messsages change they can be updated in one place instead of all
over.

Split out some new `eStrKey` values that are needed for libyang3
due to different messages, and document the libyang3 values needed once
upgraded but keep libyang1 values for now.

Finally, some tests were using both `eStr` and `eStrKey` and the `eStr`
portion is not needed to perform a meaningful validation, plus the
`eStr` in these cases are not valid for libyang3.

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
@ganglyu @qiluo-msft please review ... this needs to be merged before I can submit PRs for the migration of the libyang 1.0.73 code to v3 ...

wen587 commented:
Hi @bradh352 , could you help me understand why `src/sonic-yang-models/tests/yang_model_tests/tests/` is impacted? Aren't they dependent on libyang 1.0.73?

ganglyu commented:
@bradh352 

Thank you very much for your contribution!

We have encountered numerous issues with libyang 1.0.73, which is currently used by sonic-yang-models. On the other hand, libyang 1.0.184 and libyang 2.1.148, which are used by FRR, have not presented any issues to our knowledge.

Could we update only libyang 1.0.73 in this PR?

ganglyu commented on src/sonic-yang-models/tests/yang_model_tests/tests/asic-sensors.json:
```diff
@@ -4,7 +4,7 @@
     "ASIC_SENSORS_INVALID_POLLER_INTERVAL": {
         "desc": "Configure an invalid ASIC Sensors polling interval",
-        "eStrKey" : "Pattern"
+        "eStrKey" : "Range"
```
Why do we need to modify the unit test?

bradh352 replied:
From the commit message:
> sonic-yang-models: Update test cases in preparation for libyang3
Convert some yang model test cases that use `eStr` for failure matching
to use `eStrKey` instead to standardize expected failure matching so
when messsages change they can be updated in one place instead of all
over.
>
> Also, split out some new `eStrKey` values that are needed for libyang3
due to different messages, and document the libyang3 values needed once
upgraded but keep libyang1 values for now.
>
> Finally, some tests were using both `eStr` and `eStrKey` and the `eStr`
portion is not needed to perform a meaningful validation, plus the
`eStr` in these cases are not valid for libyang3.

So basically, its just to make the review of the upgrade to libyang3 easier.   I've actually got libyang3 working and passing these test cases so I wanted a bulk of the test case changes done here to make it easier to review the more complex porting of the libyang1 python swig module to libyang3 python cffi.

bradh352 commented:
@ganglyu 
> @bradh352
> 
> Thank you very much for your contribution!
> 
> We have encountered numerous issues with libyang 1.0.73, which is currently used by sonic-yang-models. On the other hand, libyang 1.0.184 and libyang 2.1.148, which are used by FRR, have not presented any issues to our knowledge.
> 
> My suggestion is to install libyang3 in this PR without replacing the existing dependency.

I honestly believe not having multiple versions of libyang to be extremely important.  I think their API is stabilizing.  The only thing required was to apply some commits already upstream in FRR 10.2 to support 3.7.8.  From the commit message:
> This also backports a series of commits from FRR to support libyang3,
upgrading FRR itself isn't currently in scope:
>
> * https://github.com/FRRouting/frr/commit/290c6e3184a8f6c4c5c39fc45dcec81508e3858b
> * https://github.com/FRRouting/frr/commit/8e2b2caf6592b587749777f00778bf48fa164ed5
> * https://github.com/FRRouting/frr/commit/22eccbfab916b2c6b64037e872a5c5a870d1a4d6
> * https://github.com/FRRouting/frr/commit/87c9060f90b6b9b5a69ad1e6ba6e8ebc58d94374

I think we'd be foolish to kick the can down the road while we're working on it now.   I also haven't verified if libyang2 and libyang3 can be installed simultaneously like libyang1 and libyang3 can.

Of course if you tell me there's no way you could possibly accept this PR because of this you leave me little choice but I don't see a sufficient argument at this time against it.

bradh352 commented:
> Hi @bradh352 , could you help me understand why `src/sonic-yang-models/tests/yang_model_tests/tests/` is impacted? Aren't they dependent on libyang 1.0.73?

From the commit message:
> sonic-yang-models: Update test cases in preparation for libyang3
Convert some yang model test cases that use `eStr` for failure matching
to use `eStrKey` instead to standardize expected failure matching so
when messsages change they can be updated in one place instead of all
over.
>
> Also, split out some new `eStrKey` values that are needed for libyang3
due to different messages, and document the libyang3 values needed once
upgraded but keep libyang1 values for now.
>
> Finally, some tests were using both `eStr` and `eStrKey` and the `eStr`
portion is not needed to perform a meaningful validation, plus the
`eStr` in these cases are not valid for libyang3.

So basically, its just to make the review of the upgrade to libyang3 easier.   I've actually got libyang3 working and passing these test cases so I wanted a bulk of the test case changes done here to make it easier to review the more complex porting of the libyang1 python swig module to libyang3 python cffi.

bradh352 commented:
Guess I need to figure out python wheel packaging for cross compilation to support marvell armhf (its an architecture-specific package since it uses cffi)?  I clearly am not testing such a use case currently in my local environment.  Anyone have any pointers for this?

ganglyu commented:
> @ganglyu
> 
> > @bradh352
> > Thank you very much for your contribution!
> > We have encountered numerous issues with libyang 1.0.73, which is currently used by sonic-yang-models. On the other hand, libyang 1.0.184 and libyang 2.1.148, which are used by FRR, have not presented any issues to our knowledge.
> > My suggestion is to install libyang3 in this PR without replacing the existing dependency.
> 
> I honestly believe not having multiple versions of libyang to be extremely important. I think their API is stabilizing. The only thing required was to apply some commits already upstream in FRR 10.2 to support 3.7.8. From the commit message:
> 
> > This also backports a series of commits from FRR to support libyang3,
> > upgrading FRR itself isn't currently in scope:
> > 
> > * [FRRouting/frr@290c6e3](https://github.com/FRRouting/frr/commit/290c6e3184a8f6c4c5c39fc45dcec81508e3858b)
> > * [FRRouting/frr@8e2b2ca](https://github.com/FRRouting/frr/commit/8e2b2caf6592b587749777f00778bf48fa164ed5)
> > * [FRRouting/frr@22eccbf](https://github.com/FRRouting/frr/commit/22eccbfab916b2c6b64037e872a5c5a870d1a4d6)
> > * [FRRouting/frr@87c9060](https://github.com/FRRouting/frr/commit/87c9060f90b6b9b5a69ad1e6ba6e8ebc58d94374)
> 
> I think we'd be foolish to kick the can down the road while we're working on it now. I also haven't verified if libyang2 and libyang3 can be installed simultaneously like libyang1 and libyang3 can.
> 
> Of course if you tell me there's no way you could possibly accept this PR because of this you leave me little choice but I don't see a sufficient argument at this time against it.

I will engage the owner of SONiC FRR to review the related changes.

ganglyu commented:
> Guess I need to figure out python wheel packaging for cross compilation to support marvell armhf (its an architecture-specific package since it uses cffi)? I clearly am not testing such a use case currently in my local environment. Anyone have any pointers for this?

I'm not an expert, but I believe the Sonic build pipeline uses the armhf system to build the Sonic image for Marvell armhf. Therefore, we shouldn't encounter any cross-compilation issues in the Sonic build pipeline.

bradh352 commented:
> Guess I need to figure out python wheel packaging for cross compilation to support marvell armhf (its an architecture-specific package since it uses cffi)? I clearly am not testing such a use case currently in my local environment. Anyone have any pointers for this?

Looks like azure isn't actually cross compiling at all, its running an armhf image that is advertising armv8l support which I think just means the kernel knows its a 64bit cpu ... but the system was compiled for armv7l/armhf, so python is I think determining the system based on `uname -m`.

I tried using a 64bit arm vm on my macbook to build sonic, but it immediately fails as it thinks I need to use QEMU multiarch ... even hacking the makefile to disable that and it fails later on.  Parallels doesn't allow me to directly install an armhf image, so I guess I'm dead in the water trying to test this on my own hardware right now.  I might need to pick up an rpi5 or something to build but that seems like it might be really really slow....

So I'll probably try to throw in some test commits to see if a simple 'setarch armhf' works around the issue.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

ganglyu commented on rules/libyang3-py3.mk:
```diff
@@ -0,0 +1,14 @@
+LIBYANG3_PY3_VERSION = 3.0.3
+LIBYANG3_PY3_ABI := cp$(shell $(SETARCH) python3 -c "import sysconfig; print(sysconfig.get_python_version().replace('.',''))")
+LIBYANG3_PY3_PLAT := $(shell $(SETARCH) python3 -c "import sysconfig; print(sysconfig.get_platform().replace('.','_').replace('-','_'))")
+LIBYANG3_PY3 := libyang-$(LIBYANG3_PY3_VERSION)-$(LIBYANG3_PY3_ABI)-$(LIBYANG3_PY3_ABI)-$(LIBYANG3_PY3_PLAT).whl
```
Why is LIBYANG3_PY3_PLAT necessary for Python wheels?
For instance, with swsssdk-2.0.1-py3-none-any.whl, we don't specify a platform for Python wheels.

bradh352 replied:
because this is a cffi python module meaning there is a platform-specific .so generated since it needs bindings to a C library.  The alternative would be to generate a platform-specific .deb with the bindings.  The wheel was just the easiest way ... but considering the armhf bindings it may be better to try to generate a .deb.

I can look into that conversion.  I looked at using the python setuputils stdeb stuff but I don't think that will work.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
converted to deb instead of wheel, hopefully it will fix the armhf issues

bradh352 commented:
awesome, building it as a deb works for armhf.  

just realized i accidentally removed the libyang dbg packages when i was reviewing and rewording the individual commits, going to fix that but everything should build and test successfully in this PR.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
looks like it was a spurious failure, a simple rebuild with no code changes and its fine.

bradh352 commented:
@ganglyu all tests are passing, please review.  It does look like its not unlikely to get merge conflicts because of the test case changes so it'd be great to know how long the review process might take ... there's already one from a new merge to master from today it appears.

Also, I have the next set of changes ready to go unless you want them merged into this PR.  They only port the ***test cases*** in *sonic-yang-models* to libyang3-py3, so there are no downstream dependency issues:
https://github.com/bradh352/sonic-buildimage/compare/852d821...f171545

saiarcot895 commented on files/build/versions/default/versions-web:
Typically, files within `files/build/versions` directory are not manually modified; they're instead updated by an automation. Even without these files being updated, the build should work. Did it break without these changes?

bradh352 replied:
no, i didn't check.  I can remove those changes.  I didn't look at the dependency tree on how they're used.

bradh352 replied:
done

saiarcot895 commented on src/libyang3-py3.patch/0001-debian.patch:
```diff
@@ -0,0 +1,68 @@
+--- libyang-python.orig/debian/control	1970-01-01 00:00:00.000000000 +0000
++++ libyang-python.new/debian/control	2025-02-09 12:45:51.113010546 +0000
+@@ -0,0 +1,16 @@
++Source: libyang
```
Nitpick: Can you have the source package name be `libyang-python` instead of the existing `libyang` source package name? There won't be any functional impact, it just makes it clear that the sources for the two packages are different.

(I do wish Debian packaged the python bindings as well.)

bradh352 replied:
will do, no problem

bradh352 replied:
done

saiarcot895 commented on src/libyang3-py3.patch/0001-debian.patch:
```diff
@@ -0,0 +1,68 @@
++Section: python
++Priority: optional
++Build-Depends: dh-python, python3-setuptools, python3-all-dev, python3-cffi, debhelper (>= 9), libyang-dev ( >= 3.0.3)
++Depends: libyang3
```
`Depends` should not exist in this section, please remove this line.

bradh352 replied:
ok

bradh352 replied:
done

saiarcot895 commented on src/libyang3-py3.patch/0001-debian.patch:
```diff
@@ -0,0 +1,68 @@
++
+diff -ruN libyang-python.orig/debian/files libyang-python.new/debian/files
+--- libyang-python.orig/debian/files	1970-01-01 00:00:00.000000000 +0000
++++ libyang-python.new/debian/files	2025-02-09 12:45:51.113010546 +0000
```
This file isn't needed, please remove this.

bradh352 replied:
done

bradh352 commented on src/libyang3-py3.patch/0001-debian.patch:
```diff
@@ -0,0 +1,68 @@
+--- libyang-python.orig/debian/compat	1970-01-01 00:00:00.000000000 +0000
++++ libyang-python.new/debian/compat	2025-02-09 12:45:51.113010546 +0000
+@@ -0,0 +1 @@
++9
```
sure, it was originally created with stdeb which added that by default. i'll make that change

bradh352 replied:
done

saiarcot895 commented on src/libyang3/Makefile:
```diff
@@ -2,38 +2,37 @@
-	pushd libyang2-$(LIBYANG2_VERSION)
-	#The package libyang2.1.148 is taken from debian trixie, which only has dpkg-dev version 1.21.22
+	pushd libyang-$(LIBYANG3_VERSION)
+	#The package libyang3.7.8 is taken from debian trixie, which only has dpkg-dev version 1.21.22
 	#The bullseye package has dpkg-dev version 1.20.13
 	#The VS package has dpkg-dev version 1.19.8
 	sed -i 's/dpkg-dev (>= 1.22.5)/dpkg-dev (>= 1.19.8)/' debian/control
```
Can you remove these 4 lines, as they're no longer valid? (or at last modify line 23 to remove the parts after `which only`)

bradh352 replied:
done

bradh352 committed:
libyang3: base package upgrade from libyang2

This does the base upgrade from libyang2 to libyang3.   It also drops
the completely unused libyang1 build (as the version used is in the libyang
directory which is 1.0.73).

NOTE: libyang3 depends on pkgconf, which supersedes pkg-config so there is
some churn to docker image building switching packages but it works for
all packages as a replacement.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
@saiarcot895 any other feedback after those changes?

Also any thoughts on potentially pulling in the sonic-yang-models test case migration to libyang3-py3 I have pending here: https://github.com/bradh352/sonic-buildimage/compare/852d821...f171545   ?

qiluo-msft commented:
@bradh352 seems like there will be many PRs for the goal. Could you just submit all PR (ok to keep in draft mode), and capture them in a table in this PR description?

bradh352 commented:
@qiluo-msft 
> @bradh352 seems like there will be many PRs for the goal. Could you just submit all PR (ok to keep in draft mode), and capture them in a table in this PR description?

Is there any way to have a discussion with someone that's an actual maintainer or manager so I can understand the process needed to actually accomplish a  cross repository series of PR merges?  

I'm worried about the effort involved in coordination.   I think I'll submit this PR as the only "non-breaking" PR.  Any other PRs, however, will break building and testing until ***all*** PRs across the various projects are merged.  It also means that none of the PRs will pass any tests since there are all cross-dependent.

Part of the issue is the sonic buildsystem itself, where dependencies are tracked in sonic-buildimage for subprojects like swss-common, sonic-utilities, etc.

Then another issue is sonic-utilities uses sonic-yang-mgmt, but there wasn't enough abstraction and they actually depend on libyang v1 directly, so once sonic-yang-mgmt is updated, sonic-utilities break.

Its also not possible to piece-meal it in small chunks because libyang3-dev and libyang(1)-dev conflict, its only working to build right now because libyang(1)-dev is installed later and everything that gets built later is still depending on that.  I don't see a way to have it properly uninstall and reinstall the dev packages as needed as its using Make dependencies rather than some more sophisticated build system ... and once the dependency on the "-install" is satisified, if you do the `${*}_UNINSTALLS` to force uninstallation the dependency is still satisfied so it won't be installed again.

I really don't want this to get hung up and never merged due to these complexities.

qiluo-msft commented:
Could you just submit all PR (ok to keep in draft mode), and capture them in a table in this PR description? -> This is just for code review purpose.

bradh352 commented:
> Could you just submit all PR (ok to keep in draft mode), and capture them in a table in this PR description? -> This is just for code review purpose.

I'm honestly not sure how to submit another sonic-buildimage PR that *depends* on this PR.  The patches won't merge cleanly as they're modifying some of the same files.  I can create PRs in other repos, though.

saiarcot895 commented:
> @saiarcot895 any other feedback after those changes?

Nothing else from me from a packaging perspective.

> I'm honestly not sure how to submit another sonic-buildimage PR that _depends_ on this PR. The patches won't merge cleanly as they're modifying some of the same files. I can create PRs in other repos, though.

One option is to include the changes in this PR into that PR. It will make that PR look bigger, but at least it may be easier to follow along.

bradh352 commented:
> One option is to include the changes in this PR into that PR. It will make that PR look bigger, but at least it may be easier to follow along.

Thanks, may give that a shot

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
I think this one needs to be merged sooner rather than later, i keep having to resolve merge conflicts.  I'm working on submitting the follow-up PRs

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
@qiluo-msft is the tracking table in this ticket what you are wanting?  Is my approach acceptable where I'm including a squashed commit of each dependent PR into the new PRs?

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
I'm working with the libyang guys on upstreaming the patches needed for SONiC, they've requested a slight change to the json datatype validation patch I supplied so this will be updated once I've got something they approve of.

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 committed:
libyang3-py3: patch for backlinks support

https://github.com/CESNET/libyang-python/pull/132

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
ok, rebased against latest libyang PRs after talking with the libyang guys.  I think at least some form of the patches here will be upstreamed.

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
things are being accepted upstream in libyang, so merged in the changes made there as they had slight differences in naming.  I expect upstream to have everything merged this week ... I hope...

bradh352 commented:
@ganglyu now that all PRs have been submitted in this series, can we start getting some eyes on this and get things moving?

This PR and the follow-up PR #21719 are low risk as they don't actually touch any code (other than backporting some FRR patches that are already upstream).

ganglyu commented:
Let's wait for the check pipeline to complete.

bradh352 commented:
> Let's wait for the check pipeline to complete.

It failed due to an issue in make configure unrelated to this patch.  I'll rebase it so it forces a rebuild.

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
@ganglyu 
the t1-lag test failed on both step1 and step2 PRs due to timeout, all other tests failed ... this doesn't appear in any way related to this patch (don't see how it could).  going to rebase again to force another build, but the automated test system has been really flaky lately in general.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
@ganglyu tests are failing on this PR and others due to azure pipelines issues unrelated to any actual changes.  Can we start reviewing this please?

bradh352 commented on rules/frr.mk:
```diff
@@ -8,9 +8,8 @@ export FRR_VERSION FRR_SUBVERSION FRR_BRANCH FRR_TAG
-$(FRR)_DEPENDS += $(LIBSNMP_DEV) $(LIBYANG2) $(LIBYANG2_DEV)
-$(FRR)_RDEPENDS += $(LIBYANG2)
-$(FRR)_UNINSTALLS = $(LIBYANG2_DEV) $(LIBYANG2)
+$(FRR)_DEPENDS += $(LIBSNMP_DEV) $(LIBYANG3) $(LIBYANG3_DEV)
```
when it used libyang2 it specified both so I kept that.  But yes, you're right.  I've gone ahead and removed the redundant dependency.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
any chance we can start reviewing this?

ganglyu commented:
@bradh352 you can ping the reviewers

bradh352 commented:
@saiarcot895, @k-v1, @qiluo-msft, @xumia, @lguohan, @wen587, @StormLiangMS, @liushilongbuaa

Please start reviewing this patch series on libyang3 migration.

wen587 reviewed (commented):
Yang related test change looks good to me

bhouse-nexthop commented on rules/libyang3-py3.dep:
```diff
@@ -0,0 +1,12 @@
+SPATH       := $($(LIBYANG3_PY3)_SRC_PATH)
+DEP_FILES   := $(SONIC_COMMON_FILES_LIST) rules/libyang3-py3.mk rules/libyang3-py3.dep
+DEP_FILES   += $(SONIC_COMMON_BASE_FILES_LIST)
+SMDEP_FILES := $(addprefix $(SPATH)/,$(shell cd $(SPATH) && find . -type f -exec sh -c 'git ls-files --error-unmatch "$0" >/dev/null 2>&1' {} \; -printf '%P\n'))
```
Sure, if submodules to external entities like that are frowned upon I can definitely rework it to have a custom makefile that then does the necessary patch applying and dpkg build steps.

The reason I went this way was it seemed like there was more code re-use rather than doing similar things in a bunch of Makefiles all over.

bradh352 replied:
done

bhouse-nexthop commented on src/libyang3/Makefile:
```diff
@@ -2,38 +2,36 @@
 	wget -NO "$(DSC_FILE)" $(DSC_FILE_URL)
 	wget -NO "$(ORIG_FILE)" $(ORIG_FILE_URL)
 	wget -NO "$(DEBIAN_FILE)" $(DEBIAN_FILE_URL)
-	dpkg-source -x libyang2_$(LIBYANG2_FULLVERSION).dsc
```
ok, easy enough

bradh352 replied:
done

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

wen587 commented:
Hi @xincunli-sonic , I remmeber you mentioned that there is place using libyang2. Could you help review here?

bradh352 commented:
> Hi @xincunli-sonic , I remmeber you mentioned that there is place using libyang2. Could you help review here?

FRR used to use libyang2. A full grep of a recursive checkout didn't reveal anything else, which should at least reveal any dependencies even though some things are fetched dynamically during the build.  

I've built sonic-broadcom.bin and sonic-mellanox.bin using my own tree that contains all patches applied and booted on physical hardware.

The only component I can think of that wouldn't be visible in a recursive checkout is sonic-mgmt tests.

bradh352 commented on src/libyang3-py3/Makefile:
```diff
@@ -0,0 +1,24 @@
+$(addprefix $(DEST)/, $(MAIN_TARGET)): $(DEST)/% :
+	rm -rf ./libyang-python
+	# Obtain libyang-python
+	git clone --depth=1 -b v3.0.3 https://github.com/CESNET/libyang-python.git libyang-python
```
done.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
@adyeung as per call, please escalate

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

qiluo-msft commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
@qiluo-msft I updated the first libyang3 pr, it should build and pass all tests now.  This should be able to be merged without any potential risk since the only thing using it is FRR, which already has native libyang3 support.

bradh352 committed:
buster: fix libyang3 building

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
Looks like buster failed to build due to an old version of libxxhash ... I've committed a fix for that.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 committed:
libyang3: add patch to restore LYD_VALIDATE_NOEXTDEPS

This patch was rejected upstream as it is considered a bad practice.
However, in the sonic-mgmt-common code, it has chosen to bypass
using libyang for a lot of validation, so this flag is necessary
in this use case.  sonic-mgmt-common will likely switch to goyang
in the future.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

bradh352 commented:
I keep getting spurious test failures that aren't related to this PR.  I guess I'll just keep on forcing rebuilds all tests pass.

liushilongbuaa commented:
/azpw ms_conflict

bradh352 commented:
@qiluo-msft this is merge-ready.  You approved in the past but I had upstream a fix to buster builds (not sure why buster is even built as it is EOL).  Since this is targeted for 202511 and there are 17 other PRs that depend on this to complete the libyang3 migration it is imperative this be merged sooner rather than later if that goal is going to be accomplished.

bhouse-nexthop commented on rules/libyang3.mk:
```diff
@@ -0,0 +1,28 @@
+$(LIBYANG3)_SRC_PATH = $(SRC_PATH)/libyang3
+SONIC_MAKE_DEBS += $(LIBYANG3)
+
+LIBYANG3_DEV = libyang-dev_$(LIBYANG3_FULLVERSION)_$(CONFIGURED_ARCH).deb
```
What convention are you referring to?  The package is from upstream debian, so they set the naming convention for the generate .debs, so this variable simply sets the appropriate name.

bhouse-nexthop replied:
> See also https://packages.debian.org/source/testing/libyang, where Debian chose the name `libyang-dev` for the dev header package.
> 
> Note also that this is because they only support libyang 3 going forward in Trixie, so they don't need to consider providing support for libyang 1 and 2 headers and libraries.

well that and they conflict so both can't be installed.

theasianpianist replied:
Ah typically the "libxxx" portion of the package name stays consistent between the library, header, and debug packages (so I was expecting "libyang3-dev..."). I didn't realize they aren't doing this anymore, my mistake.

kkosciuszkiewicz-arista commented on rules/libyang3-py3.dep:
```diff
@@ -0,0 +1,11 @@
+SPATH       := $($(LIBYANG3_PY3)_SRC_PATH)
+DEP_FILES   := $(SONIC_COMMON_FILES_LIST) rules/libyang3-py3.mk rules/libyang3-py3.dep
+DEP_FILES   += $(SONIC_COMMON_BASE_FILES_LIST)
+SMDEP_FILES := $(addprefix $(SPATH)/,$(shell cd $(SPATH) && find . -type f -exec sh -c 'git ls-files --error-unmatch "$0" >/dev/null 2>&1' {} \; -printf '%P\n'))
```
What is the reason for using `find . -exec 'git ls-files --error-unmatch'` for excluding the files that are not tracked by git?
Could we use `git ls-files` to achieve the same goal using a single subprocess?

bradh352 replied:
honestly I didn't evaluate the smdep_files logic at all, I just ripped it from another .dep and it worked the way I needed it to.

kkosciuszkiewicz-arista replied:
Raised #27786 to address this.
